smif/convert/__init__.py

```diff
@@ -46,230 +34,90 @@ class SpaceTimeConvertor(object):
 
     """
 
-    def __init__(self, data,
-                 from_spatial, to_spatial,
-                 from_temporal, to_temporal,
-                 region_register, interval_register):
+    def __init__(self, region_register, interval_register):
         self.logger = logging.getLogger(__name__)
-
-        self._check_uniform_units(data)
-
-        self.data = data
-        self.from_spatial = from_spatial
-        self.to_spatial = to_spatial
-        self.from_temporal = from_temporal
-        self.to_temporal = to_temporal
-
         self.regions = region_register
         self.intervals = interval_register
 
-    @property
-    def data_regions(self):
-        data_by_regions = self._regionalise_data(self.data)
-        return set(data_by_regions.keys())
-
-    @property
-    def data_by_region(self):
-        return self._regionalise_data(self.data)
-
-    def _check_uniform_units(self, data):
-        units = []
-        for entry in data:
-            units.append(entry.units)
-        if len(set(units)) > 1:
-            msg = "SpaceTimeConvertor cannot handle multiple units for conversion"
-            raise NotImplementedError(msg)
-
-    @staticmethod
-    def _regionalise_data(data):
-        """
+    def convert(self, data, from_spatial, to_spatial, from_temporal, to_temporal):
```

Comment:
(read after comments below) Thinking through optimisations/harmonisations in this module, the conversions in area and intervals could both be brought to this module and performed as a 2-dim matrix operation.  This requires coefficient matrices for region and intervals to be generated either together or independently (probably independently and then pre-combined at run-time, as not every combination of regional and temporal conversion will ever be required).  Probably a candidate for memoization.